orchestrator/modules/operators/console_output.py

```diff
@@ -0,0 +1,376 @@
+    Dynamically determines columns shown based on terminal width. Can limit number of rows.
+
+    Args:
+        discovery_space (DiscoverySpace): Source of measurement/result timeseries.
+        operation_id (str): Operation ID/tag to extract measurements for.
+        row_limit (int|None): Maximum number of table rows to show (None = no limit).
+
+    Returns:
+        Table: Rendered rich.Table object containing formatted measurement data.
+    """
+    import numpy as np
+    import pandas as pd
+
+    df: pd.DataFrame = (
+        discovery_space.complete_measurement_request_with_results_timeseries(
+            operation_id=operation_id,
+            output_format="target",
+        )
+    )
+    table_title = (
+        f"Latest measurements - {operation_id}"
+        if row_limit
+        else f"Measurements - {operation_id}"
+    )
+    table = Table(title=table_title)
+    if df.empty:
+        return table
+    # Remove unnecessary columns and setup index
+    df = df.drop(
+        columns=["result_index", "generatorid", "identifier"],
+        errors="ignore",
+    )
+    df.insert(0, "index", np.arange(len(df)))
+    # Optionally drop/convert experiment column
+    if len(discovery_space.measurementSpace.experiments) == 1:
+        df = df.drop(columns=["experiment_id"], errors="ignore")
+    else:
+        df["experiment_id"] = df["experiment_id"].apply(
+            lambda x: x.experimentIdentifier
+        )
+    # Show only as many columns as fit on screen
+    console = Console()
+    terminal_width = console.width
+    min_col_width = 12  # Minimum width per column (estimate)
+    max_columns = max(1, terminal_width // min_col_width)
+    visible_columns = list(df.columns[:max_columns])
+    hidden_columns = len(df.columns) - max_columns
+    if hidden_columns > 0:
+        visible_columns.append(f"... (+{hidden_columns} more)")
+    for col in visible_columns:
+        table.add_column(col, overflow="fold")
+    for row_number, (_, row) in enumerate(df[::-1].iterrows()):
+        if row_limit and row_number == row_limit:
+            break
+        row_data = [
+            (f"{row[col]:.2f}" if isinstance(row[col], float) else str(row[col]))
+            for col in df.columns[:max_columns]
+        ]
+        if hidden_columns > 0:
+            row_data.append("...")
+        table.add_row(*row_data)
+    return table
+
+
+def render_group_table(progress_items: dict):
+    """
+    Render a rich.Table containing all currently active spinner/progress renderables as rows.
+    Used to visually group multiple progress/spinner items in the UI.
+
+    Args:
+        progress_items (dict): Mapping from string id to Progress or Spinner objects.
+    Returns:
+        Table: Table with renderable rows for each live progress/spinner.
+    """
+    group_table = Table(title="", show_lines=False, show_edge=False, show_header=False)
+    for iid, obj in progress_items.items():
+        group_table.add_row(obj)
+    return group_table
```

Comment:
Rich also has the `Group` class for this: https://rich.readthedocs.io/en/stable/group.html